plugin/pkg/admission/noderestriction/admission.go

```diff
@@ -603,6 +640,71 @@ func (p *Plugin) admitServiceAccount(nodeName string, a admission.Attributes) er
 	return nil
 }
 
+func (p *Plugin) podReferencesAudience(pod *v1.Pod, audience string) bool {
+	for _, v := range pod.Spec.Volumes {
+		if v.Projected != nil {
+			for _, src := range v.Projected.Sources {
+				if src.ServiceAccountToken != nil && src.ServiceAccountToken.Audience == audience {
+					return true
+				}
+			}
+		}
+
+		// also allow audiences for CSI token requests
+		// - pod --> ephemeral --> pvc --> pv --> csi --> driver --> tokenrequest with audience
+		// - pod --> pvc --> pv --> csi --> driver --> tokenrequest with audience
+		// - pod --> csi --> driver --> tokenrequest with audience
+		var driverName string
+		var err error
+		switch {
+		case v.Ephemeral != nil && v.Ephemeral.VolumeClaimTemplate != nil:
+			pvcName := ephemeral.VolumeClaimName(pod, &v)
+			driverName, err = getCSIFromPVC(p.pvcGetter, p.pvGetter, pod.Namespace, pvcName)
+		case v.PersistentVolumeClaim != nil:
+			driverName, err = getCSIFromPVC(p.pvcGetter, p.pvGetter, pod.Namespace, v.PersistentVolumeClaim.ClaimName)
+		case v.CSI != nil:
+			driverName = v.CSI.Driver
+		}
+
+		if err == nil && len(driverName) > 0 {
+			if hasAudience, hasAudienceErr := csiDriverHasAudience(p.csiDriverGetter, driverName, audience); hasAudienceErr == nil && hasAudience {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// getCSIFromPVC returns the CSI driver name from the PVC->PV->CSI->Driver chain
+func getCSIFromPVC(pvcGetter corev1lister.PersistentVolumeClaimLister, pvGetter corev1lister.PersistentVolumeLister, namespace, claimName string) (string, error) {
```

Comment:
Returning `error` as well but we don't really use it anywhere. We want to check all possible options for existence of `audience` instead of fail-fast. LMK if you want me to drop this.